src/test/fake-cdn-plugin.ts

```diff
@@ -0,0 +1,144 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import type {TestRunnerPlugin} from '@web/test-runner-core/dist/server/TestRunnerPlugin.js';
+import semver from 'semver';
+
+/**
+ * A plugin for @web/test-runner that simulates an NPM CDN like unpkg.com. It
+ * extends the built-in @web/test-runner server, and can be re-configured from
+ * the test suite itself between each test using a server command
+ * (https://modern-web.dev/docs/test-runner/commands/).
+ *
+ * Usage:
+ *
+ *    import {executeServerCommand} from '@web/test-runner-commands';
+ *
+ *    const cdnData = {
+ *      "foo": {
+ *        "1.2.3": {
+ *          "main": "lib/index.js",
+ *          "files": {
+ *            "lib/index.js": {
+ *              "content": "console.log('hello');"
+ *            }
+ *          }
+ *        }
+ *      }
+ *    };
+ *    const cdnBaseUrl = await executeServerCommand('set-fake-cdn-data', cdnData);
+ *    // Redirects to <cdnBaseUrl>/foo@1.2.3/lib/index.js and serves its content.
+ *    const result = await fetch(new URL("foo@^1.0.0", cdnBaseUrl).href);
+ */
+export function fakeCdnPlugin(): TestRunnerPlugin {
+  const pathPrefix = '/fake-cdn/';
+  let baseUrl: string | undefined;
+  let data: CdnData = {};
+
+  return {
+    name: 'fake-cdn',
+
+    serverStart(args) {
+      baseUrl = `http://${args.config.hostname}:${args.config.port}${pathPrefix}`;
+    },
+
+    executeCommand({command, payload}) {
+      if (command === 'set-fake-cdn-data') {
+        data = payload as CdnData;
+        return baseUrl;
+      }
+      return false;
+    },
+
+    serve(ctx) {
+      if (!ctx.path.startsWith(pathPrefix)) {
+        return undefined;
+      }
+      const specifier = decodeURIComponent(ctx.path.slice(pathPrefix.length));
+      const parsed = parseNpmModuleSpecifier(specifier);
+      if (!parsed) {
+        ctx.response.status = 404;
+        return undefined;
+      }
+      const {pkg, semverRangeOrTag, path} = parsed;
+      const packageData = data[pkg];
+      if (!packageData) {
+        ctx.response.status = 404;
+        return undefined;
+      }
+      // Note we don't support tags other than "latest" for now, but we could
+      // easily add that to CdnData if needed.
+      const semverRange =
+        semverRangeOrTag === '' || semverRangeOrTag === 'latest'
+          ? '*'
+          : semverRangeOrTag;
+      const versions = Object.keys(packageData.versions);
+      const version = semver.maxSatisfying(versions, semverRange);
+      if (!version) {
+        ctx.response.status = 404;
+        return undefined;
+      }
+      if (version !== semverRange) {
```

Comment:
Sanity checking my understanding of the logic here. Does this mean that a concrete version has been found in `packageData.versions` that isn't for example a semverRange?

Then this is a redirect from a range to a concrete version? I.e. a range becomes resolved into something concrete such as 1.2.3. Then we redirect calling this endpoint again with the new version. Now the version 1.2.3 would match the semverRange 1.2.3 and we'd reach line 89 where we fetch the content?

Reply:
> Sanity checking my understanding of the logic here. Does this mean that a concrete version has been found in `packageData.versions` that isn't for example a semverRange?
>
> Then this is a redirect from a range to a concrete version? I.e. a range becomes resolved into something concrete such as 1.2.3. Then we redirect calling this endpoint again with the new version. Now the version 1.2.3 would match the semverRange 1.2.3 and we'd reach line 89 where we fetch the content?

Yes exactly! And we actually rely on this redirect behavior in the `CachingCdn` class, because we inspect the redirected URLs in order to generate a canonical module specifier for each import.

Added a comment to explain this too!